staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/interface.go

```diff
@@ -0,0 +1,98 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"time"
+)
+
+// QueueSetFactory knows how to make QueueSet objects.  The request
+// management filter makes a QueueSet for each priority level.  The
+// clock, and any other test-facilitating infrastructure, to use in
+// each QueueSet is known to the factory, the client does not need to
+// pass this stuff in each call to NewQueueSet.
+type QueueSetFactory interface {
+	NewQueueSet(name string, concurrencyLimit, numQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) (QueueSet, error)
+}
+
+// QueueSet is the abstraction for the queuing and dispatching
+// functionality of one non-exempt priority level.  It covers the
+// functionality described in the "Assignment to a Queue", "Queuing",
+// and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type QueueSet interface {
+	// SetConfiguration updates the configuration
+	SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) error
+
+	// Quiesce controls whether this system is quiescing.  Passing a
+	// non-nil handler means the system should become quiescent, a nil
+	// handler means the system should become non-quiescent.  A call
+	// to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.  If all the queues have no requests
+	// waiting nor executing while the system is quiescent then the
+	// handler will eventually be called with no locks held (even if
+	// the system becomes non-quiescent between the triggering state
+	// and the required call).
+	//
+	// The filter uses this for a priority level that has become
+	// undesired, setting a handler that will cause the priority level
+	// to eventually be removed from the filter if the filter still
+	// wants that.  If the filter later changes its mind and wants to
+	// preserve the priority level then the filter can use this to
+	// cancel the handler registration.
+	Quiesce(EmptyHandler)
+
+	// Wait, in the happy case, shuffle shards the given request into
+	// a queue and eventually dispatches the request from that queue.
+	// Dispatching means to return with `quiescent==false` and
+	// `execute==true`.  In one unhappy case the request is
+	// immediately rebuffed with `quiescent==true` (which tells the
+	// filter that there has been a timing splinter and the filter
+	// re-calcuates the priority level to use); in all other cases
+	// `quiescent` will be returned `false` (even if the system is
+	// quiescent by then).  In the non-quiescent unhappy cases the
+	// request is eventually rejected, which means to return with
+	// `execute=false`.  In the happy case the caller is required to
```

Comment:
I guess I expected the return signature to tell me more clearly:
* you're accepted, execute the request and call afterExecution
* The queues are too long, you're rejected
* You sat in the queue too long and timed out

I don't understand the need to tell users the queue is shutting down; IMO, remove the queueset from consideration, then shut it down. Why would a caller need to ever try again?

Reply:
The signature here takes the stance that it does not need to tell the client whether rejection is due to timeout or queues being too long or whatever; rejection is just rejection, period.  As for why quiescing shows up here, see https://github.com/kubernetes/kubernetes/pull/84544/files#r342385041 .

How about the following reworind of this comment?

> // Wait uses the given hashValue as the source of entropy as it shuffle-shards a request
> // into a queue and waits for a decision on what to do with that request.
> // If `quiescent==true` at return then the QueueSet has become undesirable and the client
> // should try to find a different QueueSet to use; `execute` and `afterExecution` are irrelevant
> // in this case.
> // Otherwise, if `execute` then the client should execute the request and then call `afterExecution()`.
> // Otherwise the client should not execute the request and `afterExecution` is irrelevant.

Reply:
* I tentatively agree it's not mandatory to expose the difference between timeout- and length-based failures.
* Even given the design (thanks for explanation in other comment) I do not understand why a caller would ever need to try again, and therefore why they would need to know the quiescent value. I think a queueset should drain completely without existing requests needing to do anything.

Reply:
The problem is that a QueueSet can become undesired between (a) the time when the client decided to use that QueueSet and (b) the time when the client makes the call to QueueSet::Wait.  It is in this case that the client needs to be told to reconsider which QueueSet to use.

Reply:
Now, you may ask, why not go ahead and put request X into that QueueSet in this case?  The answer is that the client needs a way to know when no more requests are going to get into that QueueSet, in order to be able to decide when it can finally drop the QueueSet on the floor and let GC take it.  One way to solve the problem would be with a big lock that involves both config work and request handling.  The approach taken here avoids such a big lock.

Reply:
Ah, I just asked this question in another thread, but here you've answered it.

The solution that came to mind is reference counting, which can be done via sync/atomic functions; whenever a queueset reference is about to be returned to a client, its counter is incremented. When the client is done they must call (probably via defer) a function. When it is no longer possible for the queueset to be returned as a match, and the queues are all empty, and the counter is zero, it is safe to delete.

Reply:
Also, if the big lock is a RW lock, it's not totally obvious to me that it needs to be avoided via extreme measures.

Reply:
The problem with a RW lock is that it has to be held for W while the configuration is changing, and that locks out all request processing at that time.

I do not follow the reference counting idea.  It seems to involve some sort of locking too: "whenever a queueset reference is about to be returned to a client".  Note that nothing in this layer chooses or returns a QueueSet to the client; it is the client that keeps track of his QueueSets and chooses among them.

Even though I am not sure how the reference counting would work, it looks like it will require more frequent atomic operations than the design in this PR now.

The design in this PR now does not strike me as employing extreme measures.  The request handler normally does one atomic Load to get a handle on the digested configuration, and has to try again if it runs into a concurrent modificiation of the configuration.  An optimistic fast path with retry in the rare case of concurrent interference is a common and non-drastic approach.